### PR TITLE
[WIP] handling zero dims in reindexer

### PR DESCRIPTION
**Pull request by Krovatkin**
Addresses https://github.com/pytorch/pytorch/issues/93587 by adding some extra logic to check for 0 dims

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -1252,8 +1253,21 @@ def _dynamic_reshape_indexer(old_size, new_size):
...
+            return any([size_hint(x) for x in list(l)])
+
+        if any0(old_size) or any0(new_size):
+            assert any0(old_size) and any0(new_size)
+            view_expr = [sympy.Integer(0)] * len(old_size)
+            stack_new = []
+            stack_old = []
+            numel = math.prod(vars, start=sympy.Integer(1))
+            # this is wrong as sympy_symbol is always positive
+            V.graph.sizevars.guard_equals(numel, sympy.Integer(0))
```
Comment (Krovatkin) [inline torch/_inductor/ir.py:1266]: @ngimel this guard is wrong as it's never `True` since `sympy_symbol` only creates positive symbols. I could just use `sympy.Symbol()` c-tor to create integer symbols but I wonder if this might break some invariants in inductor?

Comment (Krovatkin) [reply torch/_inductor/ir.py:1266]: @ngimel I've also tried to integrate zero-dim handling logic in the loop below, but it looked even worse than this 😢

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -1252,8 +1253,21 @@ def _dynamic_reshape_indexer(old_size, new_size):
         size_hint = V.graph.sizevars.size_hint
         vars = [sympy_symbol(f"view{i}") for i in range(len(new_size))]
 
-        stack_new = list(zip(vars, new_size))
-        stack_old = list(old_size)
+        def any0(l):
+            return any([size_hint(x) for x in list(l)])
```
Comment (ngimel) [inline torch/_inductor/ir.py:1257]: this looks like any_non_zero? On a higher level, should we get rid of zero-sized tensors before we come to actual reindexers as handling those cases here seems much trickier, as you note sympy_symbol is assumed to be positive. 
cc @chillee who last touched sympy_symbol

File: torch/_inductor/ir.py [hunk 3]
```diff
@@ -1252,8 +1253,21 @@ def _dynamic_reshape_indexer(old_size, new_size):
...
+
+        if any0(old_size) or any0(new_size):
+            assert any0(old_size) and any0(new_size)
+            view_expr = [sympy.Integer(0)] * len(old_size)
+            stack_new = []
+            stack_old = []
+            numel = math.prod(vars, start=sympy.Integer(1))
```
Comment (ngimel) [inline torch/_inductor/ir.py:1264]: I'm pretty sure this should be `sympy_product`, and `vars` is placeholder variable you just created, so you cannot guard on it, it's pointless

File: torch/_inductor/ir.py (hunk 2 above)
Comment (Chillee) [reply torch/_inductor/ir.py:1257]: Yeah, i think everything should be non-zero already?

Comment (ngimel) [reply torch/_inductor/ir.py:1257]: Apparently no, there's viewing one 0-element tensor as another one, but my suggestion is dealing with this situation before we come to reindexing.

File: torch/_inductor/ir.py (hunk 1 above)
Comment (ngimel) [reply torch/_inductor/ir.py:1266]: I think the guard should be on `sympy_product` of elements in the `old_size` to be equal to 0, and elements in the `new_size` to be equal to 0, and a special reindexer function in this case should always return 0, because that's what your offset is going to be.

File: torch/_inductor/ir.py [hunk 4]
```diff
@@ -1252,8 +1253,21 @@ def _dynamic_reshape_indexer(old_size, new_size):
         size_hint = V.graph.sizevars.size_hint
```
Comment (ngimel) [inline torch/_inductor/ir.py:1253]:
```suggestion
        old_numel = sympy_product(old_size)
        if V.graph.sizevars.maybe_guard_equals(old_numel, 0):
            new_numel = sympy_product(new_size)
            V.graph.sizevars.guard_equals(new_numel, 0)

            def reindex(index):
                return [0]*len(old_size)

            return reindex

        size_hint = V.graph.sizevars.size_hint
```

Comment (ngimel) [reply torch/_inductor/ir.py:1253]: The remaining error in trilinear test is reduction with 0 elements, should be solved by your other PR.

File: torch/_inductor/ir.py (hunk 3 above)
Comment (Krovatkin) [reply torch/_inductor/ir.py:1264]: yes, I should be guarding on the old_size vars, I mixed up the two

File: torch/_inductor/ir.py (hunk 2 above)
Comment (Krovatkin) [reply torch/_inductor/ir.py:1257]: >  On a higher level, should we get rid of zero-sized tensors before we come to actual reindexers 

@Chillee @ngimel 
could you guys elaborate? Does this mean `torchdynamo` should be graph breaking or if there should be some alternative path for lowering zero-dim tensors?

File: aten/src/ATen/core/NamedRegistrations.cpp [hunk 5]
```diff
@@ -384,7 +384,7 @@ TORCH_LIBRARY_IMPL(aten, Named, m) {
   m.impl("rand_like", CppFunction::makeFallthrough());
   m.impl("randn_like", CppFunction::makeFallthrough());
   m.impl("random_", CppFunction::makeFallthrough());
-  m.impl("random_.from", CppFunction::makeFallthrough());
+  m.impl("random_._from", CppFunction::makeFallthrough());
```
Comment (ngimel) [inline aten/src/ATen/core/NamedRegistrations.cpp:387]: why this change?
